float8_experimental/float8_linear_utils.py

```diff
@@ -120,95 +148,101 @@ def sync_float8_amax_and_scale_history(
+        Because of how this interacts with torch.compile
+
     Args:
         model (torch.nn.Module): The model to track amaxes for
-        fp8_classes (optional): The fp8 classes to look for in the model.
-            The default is Float8Linear.
-            When using with TP, users can pass in the customized TP classes instead.
         fp8_layers (optional): If fp8_layers are provided, fp8_classes are ignored,
             and we loop over all fp8_layers to sync and update amax scale histories.
             Users can use get_float8_layers to get all fp8 layers.
     """
+    if fp8_layers is None:
+        fp8_layers = get_float8_layers(model)
 
-    # For now, this is written in a naive way to maximize code readability.
-    # TODO(future): benchmark and optimize as needed, we have combined all
-    # the reductions into one and we can probably try other optimizatons to
-    # make the history update faster.
+    if len(fp8_layers) == 0:
+        log.warn(
+            "Calling sync_float8_amax_and_scale_history on a module with no Float8Linear layers"
+        )
+        return
 
-    if fp8_layers is None:
-        fp8_layers = get_float8_layers(model, fp8_classes)
+    fp8_amax_x_tensor_list = [child.fp8_amax_x for child in fp8_layers]
+    fp8_amax_w_tensor_list = [child.fp8_amax_w for child in fp8_layers]
+    fp8_amax_dL_dY_tensor_list = [child.fp8_amax_dL_dY for child in fp8_layers]
+
+    assert (
+        len(fp8_amax_x_tensor_list)
+        == len(fp8_amax_w_tensor_list)
+        == len(fp8_amax_dL_dY_tensor_list)
+    ), "Mismatched lengths of amax tensors."
 
     if dist.is_initialized():
-        fp8_amax_x_tensor = torch.tensor(
-            [child.fp8_amax_x for child in fp8_layers],
-            dtype=torch.float32,
-            device="cuda",
-            requires_grad=False,
-        )
-        fp8_amax_w_tensor = torch.tensor(
-            [child.fp8_amax_w for child in fp8_layers],
-            dtype=torch.float32,
-            device="cuda",
-            requires_grad=False,
-        )
-        fp8_amax_dL_dY_tensor = torch.tensor(
-            [child.fp8_amax_dL_dY for child in fp8_layers],
-            dtype=torch.float32,
-            device="cuda",
-            requires_grad=False,
-        )
+        # Combine all the amax tensors into one tensor and reduce it
+        fp8_amax_x_tensor = torch.cat(fp8_amax_x_tensor_list)
+        fp8_amax_w_tensor = torch.cat(fp8_amax_w_tensor_list)
+        fp8_amax_dL_dY_tensor = torch.cat(fp8_amax_dL_dY_tensor_list)
+
         dist.all_reduce(fp8_amax_x_tensor, op=dist.ReduceOp.MAX)
         dist.all_reduce(fp8_amax_w_tensor, op=dist.ReduceOp.MAX)
         dist.all_reduce(fp8_amax_dL_dY_tensor, op=dist.ReduceOp.MAX)
 
-    for idx in range(len(fp8_layers)):
-        child = fp8_layers[idx]
+        # Reassign the reduced amax values to the original tensors
+        for idx in range(len(fp8_layers)):
```

Comment:
This is the exact func I was looking for, does it not work with compile?

Reply:
I have not experimented with it much, so I think it is worth trying out.

The one time I did try it, maybe since I only compiled a small function, the compiled function was worse than eager due to some expensive setup to begin.

Reply:
Tried, out it currently does not work with AsyncTensorCollective subclass